### ISSUE TITLE
Define multi-instance functional requirements

### ISSUE THREAD

**Issue by oritstone** [openscore/score-language#147]

**Comment by meirwah** [comment]
***overview:***
MI , or multi-instance, is a way to run the same task in parallel manner.
For example : user can run a restart server on 1000 hosts simultaneously.
 
***Requirements:***
1. Ability to execute same task (either operation or subflow) in the same time.
2. MI task should be similar to regular task in the DSL structure
3. MI can iterate on a list of inputs in order to execute them in parallel manner
    1. so there is one input which is value is type list, so it can run on al values of the list to execute      them in parallel manner
    2. other inputs can be regular types (so that each task execution gets the same value for those)
4. MI task can aggregate results (declared by publish tab)
5. By default MI Task is sucess result
6. You can add a failure condition to the MI Task , so that the task will fail based on the failure of one of the task execution instances
   * need to see if you can build a rule based on each MI task instance results
 
***Open Questions:***
 
Is MI the right name? maybe go with just Parallel (unlike OO terms)?

**Comment by meirwah** [comment]
@meshipeer comments : 
my comments:
1. in case there are more than one input that is a list, which value will be taken?
2. how you navigate from it in case of more than success \ failre results
3. when publish - which order is taken? the aggregation is after all the brunches finished? each related to its own outputs?

my answers : 
1. I think we should limit support to only one , and mark it as the list to iterrate
2. limit it to only success \ failure
3. agg is after all branches finished + each related to it's own outputs

**Comment by meirwah** [comment]
adding @Bonczidai

**Comment by Bonczidai** [comment]
## DSL for MI task:

### Required task elements:

```slang
- mi_task:
        multi_instance:
          on: value in values
          do:
            ops.print:
              - text: value
              - normal_input: "'abc'"
              - flow_input
```

New keywords: `multi_instance`, `on`.

The executable under `do` will be executed for every item in values in parallel. For every execution `text` input will get the corresponding item of the list, the other inputs are treated as normal inputs (are bound from context before MI task).

### Optional task elements:

#### Publish

```slang
- mi_task:
        multi_instance:
            on: ...
            do: ...
        publish:
          - agg_output: fromInputs['agg_output'] + output + ' '
```

`publish` is on the same level with `multi_instance` suggesting aggregation only takes place after all the branches finish.
This expression means that we will iterate over all the resulting contexts from branches and evaluate the output expression. We can refer a variable from the previous branch output binding context with `fromInputs` keyword.

#### Navigate & Results

Should validate that MI executable has only SUCCESS / FAILURE results.

- case1: no navigate tag

```slang
- mi_task:
        multi_instance: ...
        publish: ...
```

MI result is SUCCESS and flow moves to the next step.

- case2: declare SUCCESS / FAILURE results

```slang
- mi_task:
        multi_instance: ...
        publish: ...
        navigate:
          SUCCESS: success_task
          FAILURE: failure_task
```

Result will be FAILURE if one of the branches fail. Here we use custom navigation.

## Affected modules

### entities

need an entity that contains `value`, `values` expression
 
### compiler

add transformer, update several classes

### runtime

need to modify task steps:
    - beginTask() will create the branches and bind inputs
    - endTask() will join the branches and aggregate the outputs


**WDYT?**

**Comment by orius123** [comment]
Few questions:
- Why do we need the `on` keyword, can't we use `for` same as loops.
- I think `async_loop` is a better name. @meirwah WDYT?
- I think that `publish` is not enough in this case. As a user I would like to have the option to run an actual script here, with a way to distinguish between branches. the other way is to have another task that handles it.
- Are the navigate rules configurable? what if I want to navigating to somewhere when 2 branches got **FAILURE**, or a **CUSTOM** result.

**Comment by meshipeer** [comment]
1. i think multi instance is not a good name name - i think we need to mention parallel somehow
2. i think the navigation is problematic, since we may have different results in different brunches
3. i think the publish is ok

**Comment by Bonczidai** [comment]
changed `on` to be `for`:

```slang
- task_name:
        parallel_loop:
          for: value in values
          do:
            ops.print:
              - text: value
              - normal_input: "'abc'"
              - flow_input
        publish:
            - agg_output: fromInputs['agg_output'] + output + ' '
        navigate:
          SUCCESS: success_task
          FAILURE: failure_task
```

Open issues:

- need a good name: parallel_loop, async_loop, parallel_split - other suggestions?
- how to publish values?
    - use suggested approach (`- agg_output: fromInputs['agg_output'] + output + ' '`)
    - add another task that handles branch contexts (and maybe navigations)?
        - however this approach gives higher flexibility this would force the user to write an operation for every parallel task
        - maybe the ability to switch between both approaches depending on complexity?
- we need to decide if we suppport other results than SUCCESS / FAILURE?
    - here we can go with simple result evaluation: if one of the branches fail -> FAILURE, otherwise SUCCESS
    - we can introduce custom rules e.g. (this may be unnecessary complication):

| syntax | semantics |
| ------------- |-------------|
| CUSTOM | all the branches end with this result |
| CUSTOM+ | at least one branch ends with this result |
| CUSTOM{2} | result appears two times | 

and so on...

I would go with:
    - `parallel_loop` for name
    - simple publish:
        - suitable for simple outputs and results
        - when we have very complex branch logic, we could publish the necessary outputs, navigate to a user defined task which would process these outputs
    - limit to SUCCESS / FAILURE results

**Comment by meirwah** [comment]
Hi,

We might need capability of each branch to publish outputs (so it can be later aggregated )
Otherwise the user cannot modify task outputs …

What do you think on the next example (created with @orius123 ):

```yaml
    - task_name:
        parallel_loop:
          for: value in values
          do:
            ops.print:
              - text: value
              - normal_input: "'abc'"
              - flow_input
          publish:
              - out1 : text[0]
              - out2 : text[1]
        aggregate:
            - first_result: ParallelBranches[0].out1
            - last_result_combined: ParallelBranches[:-1].out1 + ParallelBranches[:-1].out2
            - no_out_result: filter(lambda x: x.out1 is not None, ParallelBranches)

        navigate:
          SUCCESS: success_task
          FAILURE: failure_task
```

**Comment by meirwah** [comment]
I do not like the parallel_loop name BTW... maybe async_loop ?

**Comment by orius123** [comment]
Regarding results, I agree. SUCCESS FAILURE seems good to me.

**Comment by Bonczidai** [comment]
Hi,

please take a look at the updated DSL and technical design. I represented the flow of execution and marked where we have to change / add something:

## DSL

```
- task_name:
        async_loop:
          for: value in values
          do:
            ops.print:
              - text: value
              - normal_input: "'abc'"
              - flow_input
          publish:
              - out1 : expr1
              - out2 : expr2
        aggregate:
            - first_result: ParallelBranches[0]['out1']
            - last_result_combined: ParallelBranches[:-1]['out1'] + ParallelBranches[:-1]['out2']
            - no_out_result: filter(lambda x: x['out1'] is not None, ParallelBranches)
        navigate:
          SUCCESS: success_task
          FAILURE: failure_task
```

## Tech Design

### Entities module:

- add AsyncLoop entity: store value and value expression from `for: value in values`
- ScoreLangConstants: add async_loop constant

### Compiler module:

- SlangTextualKeys: add aggregate key
- SlangCompiler:
    - preCompile:
        - parse Yaml file
        - createModel
            - transformToExecutable
                - compileWorkflow:
                    - here we detect async_loop key, aggregate and navigate (populate actionRawData)
                    - compileTask
                        - process actionRawData
                        - run transformers: for, do, publish - no change; navigate - validate only SUCCESS / FAILURE results exists; add aggregate transformer
    - compile:
        - createFlowExecutionPlan -> buildTaskExecutionSteps: here we create begin and finish task steps and set the actionData for runtime (e.g. pass the AsyncLoop entity so at runtime we will know whether it's an MI task)
        
### Runtime module:

- add AsyncLoopBinding class:
    - evaluate input list expression
    - we can store Async Loop related information in `langVariables` in the Context object so it will be separated from the runtime variables
- TaskSteps class:
    - beginTask:
        - if async loop exists, create branches with the same execution plan ID, create branch context: `executionRuntimeServices.addBranch(stepPosition, executionPlanId, currentBranchContext);`
    - endTask:
        - join branches: `executionRuntimeServices.getFinishedChildBranchesData();`
        - publish values
        - aggregate values
        - hadle navigation

and finally a diagram of branching:

![branches](https://cloud.githubusercontent.com/assets/8058961/6662792/6c51473a-cbc8-11e4-91f4-ae6d1d16f571.png)

**Comment by Bonczidai** [comment]
We should decide which keyword to go with - `async_loop` or something else?
WDYT?

@meshipeer @oritstone  @meirwah @orius123

**Comment by orius123** [comment]
- I'm good with async_task
- you should mark step as split step in the execution plan (compiler)
- I don't think that we don't need to store the loop data in the context. engine should handle it, it is a completely different mechanism.
- at start task we should add branch for each lane.
- publish happens inside every branch. only aggregate is outside.
- maybe a better name for the ParallelBranches sequence. or enable the user to define.

**Comment by Bonczidai** [comment]
An example of a real use case from the CoreOS story (we access a list of machines based on their IDs and retrieve the IP addresses):

How it looks like with FOR:
```yaml
    - get_machine_public_ip:
            loop:
                for: machine_id in machines_id_list.split(' ')
                do:
                  coreos.get_machine_public_ip:
                    - machine_id
                    - host: coreos_host
                    - username: coreos_username
                    - password: coreos_password
                    - privateKeyFile: private_key_file
                publish:
                    - machines_public_ip_list: fromInputs['machines_public_ip_list'] + public_ip + ' '
                    - error_message
```

How it would like with MI:
```yaml
    - get_machine_public_ip:
            async_task:
                for: machine_id in machines_id_list.split(' ')
                do:
                  coreos.get_machine_public_ip:
                    - machine_id
                    - host: coreos_host
                    - username: coreos_username
                    - password: coreos_password
                    - privateKeyFile: private_key_file
                publish:
                    - public_ip
                    - error_message
            aggregate:
                    - machines_public_ip_list: map(lambda x:x['public_ip'], ParallelBranches)
                    # or if we want string output: ' '.join(map(lambda x:x['public_ip'], ParallelBranches))
                    - error_message: map(lambda x:x['error_message'], ParallelBranches)
```

**Comment by orius123** [comment]
I really like it.
Can you please explain here the publish issue that we discussed so everyone will be able to share their opinions?

**Comment by Bonczidai** [comment]
Publish issue:

Logically publish should happen for every branch so we can aggregate the published outputs in the endTask method (where we join the branches). But running publish inside branches is not straightforward since the executable (flow / operation) has its own execution plan (with outputs) that is NOT dependent on the MI task.

Currently I see three approaches:

- publish is only semantic: we do both publish and aggregate in endTask method by iterating the result context (we trick the user here)
- add a wrapper execution plan that has an additional start- and finishExecutble step which are used only for publish - the endTask method would do the aggregation (we trick ourselves here)
- use only aggregate without publish

See figure:
![branches2](https://cloud.githubusercontent.com/assets/8058961/6683576/969f1486-cc90-11e4-9119-de4b464f798d.png)

**Comment by Bonczidai** [comment]
Results:

- we will go with async_loop keyword
- we will go with Branches_Contexts in aggregate expressions
- currently if none of the branches fail, the async loop ends with SUCCESS - we will need to add in the future more complex result expressions (should add this to the **features** list)
- we will introduce two new steps for async loop: split and join steps - see figure again :)

![branches3](https://cloud.githubusercontent.com/assets/8058961/6711726/649fde64-cd91-11e4-9a42-f0061e71db20.png)

**Comment by orius123** [comment]
I think branches_context should be lowercase WDYT?
Please add future task to iterate over iterables.

**Comment by oritstone** [comment]
:+1:

**Comment by Bonczidai** [comment]
sure..

so as future tasks, we will have:

- result expression support for async loop (based on branches)
- iterate over iterables in for (not only lists)
